web/interceptor/access_control.go

```diff
@@ -0,0 +1,150 @@
+	student
+	// teacher: user enrolled in the course with teacher status.
+	teacher
+	// courseAdmin: an admin user who is also enrolled into the course.
+	courseAdmin
+	// admin is the user with admin privileges.
+	admin
+)
+
+// If there are several roles that can call a method, a role with the least privilege must come first.
+// If method is not in the map, there is no restrictions to call it.
+var access = map[string]roles{
+	"GetUser":                {},
+	"GetCourse":              {},
+	"GetCourses":             {},
+	"CreateEnrollment":       {},
+	"GetCoursesByUser":       {},
+	"GetAssignments":         {},
+	"GetRepositories":        {},
+	"UpdateCourseVisibility": {user},
+	// TODO(vera): needs a specific check: if request attempts to change admin role, user role is not sufficien.
+	"UpdateUser":              {user, admin},
+	"GetEnrollmentsByUser":    {user, admin},
+	"GetSubmissions":          {user, group, teacher, courseAdmin},
+	"GetGroupByUserAndCourse": {group, teacher},
+	"CreateGroup":             {group, teacher},
+	"GetGroup":                {group, teacher},
+	"GetEnrollmentsByCourse":  {student, teacher},
+	"UpdateGroup":             {teacher},
+	"DeleteGroup":             {teacher},
+	"GetGroupsByCourse":       {teacher},
+	"UpdateCourse":            {teacher},
+	"UpdateEnrollments":       {teacher},
+	"UpdateAssignments":       {teacher},
+	"UpdateSubmission":        {teacher},
+	"UpdateSubmissions":       {teacher},
+	"RebuildSubmissions":      {teacher},
+	"CreateBenchmark":         {teacher},
+	"UpdateBenchmark":         {teacher},
+	"DeleteBenchmark":         {teacher},
+	"CreateCriterion":         {teacher},
+	"UpdateCriterion":         {teacher},
+	"DeleteCriterion":         {teacher},
+	"CreateReview":            {teacher},
+	"UpdateReview":            {teacher},
+	"GetReviewers":            {teacher},
+	"IsEmptyRepo":             {teacher},
+	"GetSubmissionsByCourse":  {teacher, courseAdmin},
+	"GetUserByCourse":         {teacher, admin},
+	"GetUsers":                {admin},
+	"GetOrganization":         {admin},
+	"CreateCourse":            {admin},
+}
+
+func AccessControl(logger *zap.SugaredLogger, tm *auth.TokenManager) grpc.UnaryServerInterceptor {
+	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+		method := info.FullMethod[strings.LastIndex(info.FullMethod, "/")+1:]
+		logger.Debugf("ACCESS CONTROL for method %s", method) // tmp
+		roles, ok := access[method]
+		if ok {
+			logger.Debug("Got roles: ", roles) // tmp
+			claims, err := tm.GetClaims(ctx)
+			if err != nil {
+				logger.Error("Access control: failed to get claims from request context: %v", err)
+				return handler(ctx, req)
+			}
+			logger.Debug("Got user claims: ", claims) // tmp
+			for _, role := range roles {
+				switch role {
+				case user:
+					if m, ok := req.(requestID); ok {
```

Comment:
Here you can then do:
```go
if hasID(claims.UserID) {
	return handler(ctx, req)
}
```
and similar for the others that check only equality with an ID.